crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,262 @@
+It follows from this that for every sealed type,
+there is also only a finite and pre-determined set of *subtypes* of that type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes of `bool` are `Literal[True]`, `Literal[False]`, and `Never`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+        Z = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X`, `Foo.Y` and `Foo.Z`.
+
+    - The only proper subtypes of `Foo` are `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]`,
+        `Literal[Foo.X, Foo.Y]`, `Literal[Foo.X, Foo.Z]`, `Literal[Foo.Y, Foo.Z]`, and `Never`.
+
+For a given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
+`X & ~A == B | C`.
+
+Due to the enumerability of a sealed type's inhabitants,
+a Python sealed type can be narrowed using identity:
+
+```py
+from enum import Enum, auto
+
+class Colors(Enum):
+    RED = auto()
+    BLUE = auto()
+    YELLOW = auto()
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[Colors.RED, Colors.BLUE, Colors.YELLOW]`)
+
+    if var is Colors.RED:
+        ...  # var is narrowed to `Literal[Colors.RED]`
+    else:
+        ...  # var is narrowed to `Literal[Colors.BLUE, Colors.YELLOW]`
+```
+
+## Single-value types
+
+For a given type `T`, `T` can be said to be a single-value type if there exists an
+[equivalence relation](https://en.wikipedia.org/wiki/Equivalence_relation)
+between all inhabitants of the type and the equivalence relation is satisfied between
+all inhabitants of the type.
+
+For a Python type `T` to be categorised as a single-value type:
+
+- All inhabitants of `T` must be instances of the same runtime class `U`
+- `U` must have `__eq__` and `__ne__` methods such that the equality relation between instances of `U`
+    is reflexive, symmetric, and transitive.
+- For any two inhabitants of `T` named `x` and `y`, `x == y`.
+
+Nearly all singleton types are single-value types, but the reverse is not true.
+Many single-value types exist that are not singleton types: examples include
+`Literal[123456]`, `Literal[b"foo"]`, `tuple[Literal[True], Literal[False]]`, and `Literal["foo"]`.
+All runtime inhabitants of the `Literal["foo"]` type are equal to the string `"foo"`.
+However, they are not necessarily the *same* object in terms of identity;
+multiple `str` instances equal to `"foo"` can exist at runtime at different memory addresses.[^1]
+This means that it is not safe to narrow a single-value type by identity unless it is also known
+that the type is a singleton type.
```

Comment:
We said above that we can also narrow sealed types by identity, not just singleton types:
```suggestion
that the type is a sealed type.
```